modvendor/modvendor.go

```diff
@@ -0,0 +1,223 @@
+// Vendoring is not recursive, so dependencies won't have their dependencies vendored.
+// Vendoring will also not download any git submodules.
+//
+// It returns the absolute path where the code has been vendored, which will be inside
+// the given vendordir.
+func Vendor(vendordir string, src Source) (string, error) {
+	logger := log.With().
+		Str("action", "modvendor.Vendor()").
+		Str("vendordir", vendordir).
+		Str("url", src.URL).
+		Str("path", src.Path).
+		Str("ref", src.Ref).
+		Logger()
+
+	if src.Ref == "" {
+		// TODO(katcipis): handle default references.
+		// for now always explicit is fine.
+		return "", errors.E("src %v reference must be non-empty", src)
+	}
+
+	// TODO(katcipis): test that if vendor contains path with matching ref
+	// it will do nothing.
+
+	logger.Trace().Msg("setting up tmp workdir")
+
+	workdir, err := os.MkdirTemp("", "terramate-vendor")
+	if err != nil {
+		return "", errors.E(err, "creating workdir")
+	}
+	defer func() {
+		// We ignore the error here since after the final os.Rename
+		// the workdir will be moved and won't exist.
+		_ = os.Remove(workdir)
+	}()
+
+	clonedir := filepath.Join(vendordir, src.Path, src.Ref)
+
+	logger = logger.With().
+		Str("workdir", workdir).
+		Str("clonedir", clonedir).
+		Logger()
+
+	logger.Trace().Msg("setting up git wrapper")
+
+	g, err := git.WithConfig(git.Config{
+		WorkingDir: workdir,
+	})
+	if err != nil {
+		return "", err
+	}
+
+	logger.Trace().Msg("cloning to workdir")
+
+	if err := g.Clone(src.URL, workdir); err != nil {
+		return "", err
+	}
+
+	// TODO(katcipis): remove .git before moving.
+
+	if err := os.MkdirAll(filepath.Dir(clonedir), 0775); err != nil {
+		return "", errors.E(err, "creating mod dir inside vendor")
+	}
+
+	logger.Trace().Msg("moving cloned mod from workdir to clonedir")
+	if err := os.Rename(workdir, clonedir); err != nil {
+		// This may leave intermediary created dirs hanging on vendordir
+		// since we just create all and then delete clone dir on a failure to move.
+		// If we get a lot of errors from os.Rename we may need to handle this
+		// more gracefully, here we assume that os.Rename errors are rare since both
+		// dirs where just created.
+		errs := errors.L()
+		errs.Append(errors.E(err, "moving cloned module"))
+		errs.Append(os.Remove(clonedir))
+		return "", errs.AsError()
+	}
+	return clonedir, nil
+}
+
+// ParseSource parses the given modsource string.
+// The modsource must be a valid Terraform Git/Github source reference as documented in:
+//
+// - https://www.terraform.io/language/modules/sources
+//
+// Source references that are not Git/Github are not supported.
+func ParseSource(modsource string) (Source, error) {
+	ref := ""
+	splitParams := strings.Split(modsource, "?")
+	modsource = splitParams[0]
+
+	if len(splitParams) > 1 {
+		if len(splitParams) != 2 {
+			return Source{}, errors.E(ErrInvalidModSrc, "unexpected extra '?' on source")
```

Comment:
the error message is not clear what is "extra" and which ones are supported.
Maybe something like this?

```suggestion
			return Source{}, errors.E(ErrInvalidModSrc, "only `?ref=<val>` query parameter is supported but found %q", strings.Join(splitParams[1:], "?"))
```

Reply:
Actually this error indicates that you have more than one ? in the string, isn't it clear enough to say the string has more than one '?' ?

Reply:
no the error says you have "extra stuff" in the query parameter, you don't say you expect one.

Changing this to `%d extra query parameters` is better but maybe we should be already clear about what we expect there: just the `ref`...

Reply:
I was trying to be precise that there is an extra `?` but now I get that you are saying that ? makes you think it would be an extra anything ? :laughing: